src/Map.ts

```diff
@@ -0,0 +1,465 @@
+ *
+ * @category Lookups
+ *
+ * @returns An `Option` containing a tuple of the key and value.
+ */
+const findWithKey =
+    <K>(key: K, { equals }: EqualityComparer<K> = defaultEqualityComparer) =>
+    <V>(map: ReadonlyMap<K, V>): Option<[K, V]> => {
+        if (map.size < 1) {
+            return Option.none
+        }
+
+        for (const [k, v] of map) {
+            if (equals(k, key)) {
+                return Option.some([k, v])
+            }
+        }
+
+        return Option.none
+    }
+
+/**
+ * Test whether a `Map` contains the given key. Uses the given `EqualityComparer`
+ * if passed. Otherwise, defaults to reference equality (triple equals).
+ *
+ * @category Lookups
+ *
+ * @returns `true` if the key is in the `Map`, `false` otherwise.
+ */
+const containsKey =
+    <K>(key: K, equalityComparer: EqualityComparer<K> = defaultEqualityComparer) =>
+    <V>(map: ReadonlyMap<K, V>): boolean =>
+        pipe(map, findWithKey(key, equalityComparer), Option.isSome)
+
+/**
+ * Get a value associated with the given key from the `Map`. Returns a `Some`
+ * containing the value, or `None` if the key is not in the `Map`.
+ *
+ * Uses the given equality comparer if passed, otherwise defaults to using
+ * reference equality (triple equals) for equality comparison.
+ *
+ * @category Lookups
+ */
+const find =
+    <K>(key: K, equalityComparer: EqualityComparer<K> = defaultEqualityComparer) =>
+    <V extends {}>(map: ReadonlyMap<K, V>): Option<V> =>
+        pipe(
+            map,
+            findWithKey(key, equalityComparer),
+            Option.map(([, v]) => v)
+        )
+
+/**
+ * Adds a key/value pair to a `Map`. If the given key already exists
+ * the value at that key will be updated with the given value.
+ *
+ * Will use the equality comparer if given, otherwise defaults to using
+ * reference equality (triple equals) for equality comparisons.
+ *
+ * @category Transformations
+ *
+ * @returns A new `Map` with the added key/value pair
+ */
+const add =
+    <K, V>(
+        [key, value]: readonly [K, V],
+        equalityComparer: EqualityComparer<K> = defaultEqualityComparer
+    ) =>
+    (map: ReadonlyMap<K, V>): ReadonlyMap<K, V> => {
+        if (map.size < 1) {
+            const out = empty<K, V>()
+            out.set(key, value)
+            return out
+        }
+
+        const copy = new globalThis.Map(map)
+
+        return pipe(
+            map,
+            findWithKey(key, equalityComparer),
+            Option.match({
+                none: () => {
+                    copy.set(key, value)
+                    return copy
+                },
+                some: ([k]) => {
+                    copy.set(k, value)
+                    return copy
+                },
+            })
+        )
+    }
+
+/**
+ * Make a new `Map` by producing a new each value for each key using
+ * the given function.
+ *
+ * @category Mapping
+ * @category Transformations
+ */
+const map =
```

Comment:
Looking forward to using this like this: `Map.map(mapMap)(map)` :trollface:

Reply:
💯